### PR TITLE
[amp-story-player] ✨♻️  Rewind API

### LINKED ISSUE #31424: [amp-story-player] Introduce rewindStory() API
When the story player moves to a new story, I want it to start from the first page. 

So one way that was suggested was to call `player.go(0, -9999)` when a story changes. It does not work though. 
I listen to the `navigation` event and call the `go` method but nothing happens. 

See: https://jsfiddle.net/9x327jep/9/

On the first story o the second page, click on 'Third Story' to go to another story. Let it run to page 3 of this story. 
Hit 'First Story' again and 'Third Story' again. It continues from the page it left on.

### PR DESCRIPTION
**Pull request by Enriqe**
Closes #31424

Adds `player.rewind(storyUrl)` API to replay a given story inside of the player.

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story.js [hunk 1]
```diff
@@ -2652,6 +2657,8 @@ export class AmpStory extends AMP.BaseElement {
           ? NavigationDirection.NEXT
           : NavigationDirection.PREVIOUS
       );
+    } else if (data['rewind']) {
+      this.initialStoryLayoutDeferred_.promise.then(() => this.replay_());
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:2661]: Please use the signals API

```
this.signals().whenSignal(CommonSignals.LOAD_END).then(() => this.replay_());
```

File: src/amp-story-player/amp-story-player-impl.js [hunk 2]
```diff
@@ -1264,6 +1281,32 @@ export class AmpStoryPlayer {
...
+    }
+
+    let connectedPromise = Promise.resolve();
+    if (!story.iframe.isConnected) {
+      connectedPromise = story.connectedDeferred.promise;
+    }
+
+    connectedPromise.then(() => {
+      story.messagingPromise.then((messaging) =>
+        messaging.sendRequest('selectPage', {'rewind': true})
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:1305]: I think it'd make sense to create a new `rewind` or `replay` message, it does more than `selectPage`

File: src/amp-story-player/amp-story-player-impl.js [hunk 3]
```diff
@@ -1264,6 +1281,32 @@ export class AmpStoryPlayer {
...
+   */
+  rewind(storyUrl) {
+    const storyIdx = storyUrl
+      ? findIndex(this.stories_, ({href}) => href === storyUrl)
+      : this.currentIdx_;
+
+    const story = this.stories_[storyIdx];
+    if (!story) {
+      throw new Error(`Story URL not found in the player: ${storyUrl}`);
+    }
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:1296]: Let's not copy/paste this code into each single method we add, and create a method for it

**Review by gmajoulet — COMMENTED** [review]
I don't want to block the PR, but two comments for future reference (or for this PR if you want to):
- the `initializeAndAddStory_` refactoring is pretty unrelated and should have its own PR so we can really see if it makes the code easier, and not focus on different things in one PR
- do you feel strongly about supporting `rewind()` for stories that are not loaded? What happens today if I navigate to the last page of story A, navigate by a few more stories, and then come back to story A? This extra feature could be an extra PR too. The ideal workflow would be: create a ticket for the basic feature that sends a message to amp-story, create a ticket to support rewind() for non loaded stories, so we can dive into the behavior and discuss if/how to implement it

**Comment by Enriqe** [comment]
I did the split and filed https://github.com/ampproject/amphtml/issues/32941 for the follow up and created PR https://github.com/ampproject/amphtml/pull/32944 for the refactoring @gmajoulet

File: src/amp-story-player/amp-story-player-impl.js [hunk 4]
```diff
@@ -128,7 +128,8 @@ let DocumentStateTypeDef;
  *   messagingPromise: ?Promise,
  *   title: (?string),
  *   posterImage: (?string),
- *   storyContentLoaded: ?boolean
+ *   storyContentLoaded: ?boolean,
+ *   connectedDeferred: !Deferred
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:132]: We don't need the references to `connectedDeferred` before the next PR

File: src/amp-story-player/amp-story-player-impl.js [hunk 5]
```diff
@@ -1264,6 +1263,39 @@ export class AmpStoryPlayer {
...
+
+  /**
+   * Rewinds the given story.
+   * @param {string} storyUrl
+   */
+  rewind(storyUrl) {
+    const story = this.stories_[this.getStoryIdxFromUrl_(storyUrl)];
+
+    story.messagingPromise.then((messaging) =>
+      messaging.sendRequest('selectPage', {'rewind': true})
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:1295]: Re: introducing a new message, I really think a `rewind` API makes sense as it does more than just selecting a page, it's not equivalent to navigating to page 0

File: src/amp-story-player/amp-story-player-impl.js [hunk 6]
```diff
@@ -1264,6 +1263,39 @@ export class AmpStoryPlayer {
     );
   }
 
+  /**
+   * Returns the storyIdx of a story given a URL.
+   * @param {string} storyUrl
+   * @return {number}
+   * @private
+   */
+  getStoryIdxFromUrl_(storyUrl) {
```
Comment (gmajoulet) [inline src/amp-story-player/amp-story-player-impl.js:1272]: Given that we just refactored everything to only work with StoryDefs, this should return a StoryDef and not a number. If the caller happens to need the index, all it takes is `story.idx`

File: test/unit/test-amp-story-player.js [hunk 7]
```diff
@@ -799,6 +799,24 @@ describes.realWin('AmpStoryPlayer', {amp: false}, (env) => {
...
+
+      const sendRequestSpy = env.sandbox.spy(fakeMessaging, 'sendRequest');
+      player.rewind('https://example.com/story0.html');
+      await nextTick();
+
+      expect(sendRequestSpy).to.have.been.calledWith('selectPage', {
+        'rewind': true,
+      });
+    });
+
```
Comment (gmajoulet) [inline test/unit/test-amp-story-player.js:819]: Can you also add a test with a wrong URL? We won't need to add one per new public method, but just one to test the error case in getStoryFromUrl would be cool

**Review by gmajoulet — APPROVED** [review]

**Pull request merged**